src/app/chart/donut/donut.component.ts

```diff
@@ -0,0 +1,144 @@
+import { Component, DoCheck, Input, OnInit } from '@angular/core';
+
+import { cloneDeep, defaults, isEqual, merge } from 'lodash';
+import { ChartDefaults } from '../chart.defaults';
+import { ChartBase } from '../chart.base';
+import { DonutConfig } from './donut-config';
+
+import * as d3 from 'd3';
+
+@Component({
+  selector: 'pfng-chart-donut',
+  templateUrl: './donut.component.html'
+})
+export class DonutComponent extends ChartBase implements DoCheck, OnInit {
+
+  @Input() chartData: any;
+  @Input() config: DonutConfig;
+
+  public donutChartId: any;
+
+  private prevChartData: any;
+  private defaultConfig: DonutConfig;
+
+  /**
+   * Default constructor
+   * @param chartDefaults
+   */
+  constructor(private chartDefaults: ChartDefaults) {
+    super();
+  }
+
+  ngOnInit(): void {
+    this.donutChartId = 'donutChart';
+    if (this.config.chartId) {
+      this.donutChartId = this.config.chartId + this.donutChartId;
+    }
+
+    this.chartLoaded.subscribe({
+      next: (chart: any) => {
+        this.chartAvailable(chart);
+      }
+    });
+
+    this.setupConfig();
+    super.generateChart(this.donutChartId);
+    this.updateAll();
```

Comment:
You appear to be calling updateAll and generateChart twice. Once via ngDoCheck and again here. 

I believe we want to call generateChart and updateAll once, within ngDoCheck/setupConfig? That is, if and only if there is a change to chartData or config.

Reply:
This comment and the one on `updateAll` triggered a re-evaluation of the update code. For context this was originally taken from the `sparkline` component already in patternfly-ng. I have revised the design heavily and I think the original in `sparkline` component could also be revised in a similar matter. I have also included a temporary commit in the branch to demonstrate the update capabilities for both the data-only and config-only situation via the examples. Please let me know what you think.

In general I think there is still an issue that makes working with the code more difficult than necessary, however I did not want to make any changes that would affect all charts in this PR.

This is that the user inputs `config` and `data` separately in the component, while internally, these are combined under the `config` object (`config && config.data`). This ends up making updates more complicated as an update to `config` should trigger a full reload of the chart, while an update of the users `data` object only needs to trigger a reload of the data. Since they're combined, we cannot make `config.data` a reference to the users data object, as changes to the users data would result in config being not equal, triggering a full reload instead of just a data reload. At the moment, to workaround this, it's a deep clone instead (from the `merge` function), resulting in extra memory, and extra cycles used.

Reply:
It appears that merge with config.data was taken almost verbatim from APF. It was using a merge utility, but no idea why. Regardless, we're not using that and the chart doesn't appear to need it. Lets get rid of that and keep the data and config separate.